File: eprofiler-core/src/test/java/cz/encircled/eprofiler/test/ConfigTest.java
package cz.encircled.eprofiler.test;

import cz.encircled.eprofiler.core.ProfileConfiguration;
import org.junit.Assert;
import org.junit.Test;

/**
 * @author Vlad on 25-May-16.
 */
public class ConfigTest {

    @Test(expected = RuntimeException.class)
    public void testNullArgs() {
        new ProfileConfiguration(null);
    }

    @Test(expected = RuntimeException.class)
    public void testWrongArgFormat() {
        new ProfileConfiguration("classPattern=some=value;debug");
    }

    @Test(expected = RuntimeException.class)
    public void testClassPatternMissing() {
        new ProfileConfiguration("debug;");
    }

    @Test(expected = RuntimeException.class)
    public void testClassPatternNotSet() {
        new ProfileConfiguration("classPattern=");
    }

    @Test
    public void testClassPattern() {
        assertClassPatternBasicTest(new ProfileConfiguration("classPattern=cz.eprofiler"), "cz.eprofiler");

        ProfileConfiguration c2 = new ProfileConfiguration("classPattern=cz.eprofiler;");
        Assert.assertEquals(1, c2.getClassNamePatterns().size());
        Assert.assertEquals("cz.eprofiler", c2.getClassNamePatterns().get(0));

        ProfileConfiguration c3 = new ProfileConfiguration("debug;classPattern=org.some.*");
        Assert.assertEquals(1, c3.getClassNamePatterns().size());
        Assert.assertEquals("org.some.*", c3.getClassNamePatterns().get(0));

        ProfileConfiguration c4 = new ProfileConfiguration("debug;classPattern=org.some.*;");
        Assert.assertEquals(1, c4.getClassNamePatterns().size());
        Assert.assertEquals("org.some.*", c4.getClassNamePatterns().get(0));
    }

    @Test
    public void testMultipleClassPattern() {
        assertClassPatternBasicTest(new ProfileConfiguration("classPattern=cz.eprofiler;debug;classPattern=ex.ample;"), "cz.eprofiler", "ex.ample");
        assertClassPatternBasicTest(new ProfileConfiguration("classPattern=cz.eprofiler;debug;classPattern=ex.ample;"), "cz.eprofiler", "ex.ample");
        assertClassPatternBasicTest(new ProfileConfiguration("classPattern=cz.eprofiler;classPattern=ex.ample"), "cz.eprofiler", "ex.ample");
        assertClassPatternBasicTest(new ProfileConfiguration("debug;classPattern=cz.eprofiler;classPattern=ex.ample"), "cz.eprofiler", "ex.ample");
    }

    private void assertClassPatternBasicTest(ProfileConfiguration conf, String... patterns) {
        Assert.assertEquals(patterns.length, conf.getClassNamePatterns().size());
        for (int i = 0; i < patterns.length; i++) {
            Assert.assertEquals(patterns[i], conf.getClassNamePatterns().get(i));
        }
    }

    @Test
    public void testDebug() {
        Assert.assertTrue(new ProfileConfiguration("classPattern=some;debug;").isDebug());
        Assert.assertTrue(new ProfileConfiguration("classPattern=some;debug").isDebug());
        Assert.assertTrue(new ProfileConfiguration("debug;classPattern=some;").isDebug());
        Assert.assertTrue(new ProfileConfiguration("debug;classPattern=some").isDebug());
    }

    @Test
    public void testMinDuration() {
        Assert.assertEquals(10L, new ProfileConfiguration("classPattern=some;").getMinDurationToLog());
        Assert.assertEquals(11L, new ProfileConfiguration("classPattern=some;minDurationToLog=11").getMinDurationToLog());
        Assert.assertEquals(11L, new ProfileConfiguration("classPattern=some;minDurationToLog=11;").getMinDurationToLog());
        Assert.assertEquals(12L, new ProfileConfiguration("minDurationToLog=12;classPattern=some;").getMinDurationToLog());
    }

}

File: eprofiler-core/src/test/java/cz/encircled/eprofiler/test/ConfigTest.kt
package cz.encircled.eprofiler.test

import cz.encircled.eprofiler.core.ProfileConfiguration
import org.junit.Assert
import org.junit.Test

/**
 * @author Vlad on 25-May-16.
 */
class ConfigTest {

    @Test(expected = RuntimeException::class)
    fun testNullArgs() {
        ProfileConfiguration(null)
    }

    @Test(expected = RuntimeException::class)
    fun testWrongArgFormat() {
        ProfileConfiguration("classPattern=some=value;debug")
    }

    @Test(expected = RuntimeException::class)
    fun testClassPatternMissing() {
        ProfileConfiguration("debug;")
    }

    @Test(expected = RuntimeException::class)
    fun testClassPatternNotSet() {
        ProfileConfiguration("classPattern=")
    }

    @Test
    fun testClassPattern() {
        assertClassPatternBasicTest(ProfileConfiguration("classPattern=cz.eprofiler"), "cz.eprofiler")

        val c2 = ProfileConfiguration("classPattern=cz.eprofiler;")
        Assert.assertEquals(1, c2.getClassNamePatterns().size.toLong())
        Assert.assertEquals("cz.eprofiler", c2.getClassNamePatterns()[0])

        val c3 = ProfileConfiguration("debug;classPattern=org.some.*")
        Assert.assertEquals(1, c3.getClassNamePatterns().size.toLong())
        Assert.assertEquals("org.some.*", c3.getClassNamePatterns()[0])

        val c4 = ProfileConfiguration("debug;classPattern=org.some.*;")
        Assert.assertEquals(1, c4.getClassNamePatterns().size.toLong())
        Assert.assertEquals("org.some.*", c4.getClassNamePatterns()[0])
    }

    @Test
    fun testMultipleClassPattern() {
        assertClassPatternBasicTest(ProfileConfiguration("classPattern=cz.eprofiler;debug;classPattern=ex.ample;"), "cz.eprofiler", "ex.ample")
        assertClassPatternBasicTest(ProfileConfiguration("classPattern=cz.eprofiler;debug;classPattern=ex.ample;"), "cz.eprofiler", "ex.ample")
        assertClassPatternBasicTest(ProfileConfiguration("classPattern=cz.eprofiler;classPattern=ex.ample"), "cz.eprofiler", "ex.ample")
        assertClassPatternBasicTest(ProfileConfiguration("debug;classPattern=cz.eprofiler;classPattern=ex.ample"), "cz.eprofiler", "ex.ample")
    }

    private fun assertClassPatternBasicTest(conf: ProfileConfiguration, vararg patterns: String) {
        Assert.assertEquals(patterns.size.toLong(), conf.getClassNamePatterns().size.toLong())
        for (i in patterns.indices) {
            Assert.assertEquals(patterns[i], conf.getClassNamePatterns()[i])
        }
    }

    @Test
    fun testDebug() {
        Assert.assertTrue(ProfileConfiguration("classPattern=some;debug;").isDebug)
        Assert.assertTrue(ProfileConfiguration("classPattern=some;debug").isDebug)
        Assert.assertTrue(ProfileConfiguration("debug;classPattern=some;").isDebug)
        Assert.assertTrue(ProfileConfiguration("debug;classPattern=some").isDebug)
    }

    @Test
    fun testMinDuration() {
        Assert.assertEquals(10L, ProfileConfiguration("classPattern=some;").minDurationToLog)
        Assert.assertEquals(11L, ProfileConfiguration("classPattern=some;minDurationToLog=11").minDurationToLog)
        Assert.assertEquals(11L, ProfileConfiguration("classPattern=some;minDurationToLog=11;").minDurationToLog)
        Assert.assertEquals(12L, ProfileConfiguration("minDurationToLog=12;classPattern=some;").minDurationToLog)
    }

}
